### PR TITLE
method returns and return tests

### PR DESCRIPTION
**Pull request by realmarcin**

### CODE REVIEW

File: lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java [hunk 1]
```diff
@@ -117,13 +119,16 @@ public void estimateK(EstimateKParams params,
                 params.getNeighbSize(), params.getMaxItems());
         provenance.get(0).withDescription("K estimation for K-Means clustering method")
                 .withInputWsObjects(Arrays.asList(params.getInputMatrix()));
-        getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(params.getOutWorkspace())
+	WorkspaceClient wsc = getWsClient();
+        List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,Map<String,String>>> getInfo = wsc.saveObjects(new SaveObjectsParams().withWorkspace(params.getOutWorkspace())
```
Comment (MrCreosote) [inline lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java]: getInfo seems to be unused

Comment (MrCreosote) [reply lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java]: Looks like this happens in a number of places

File: lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java [hunk 2]
```diff
@@ -168,10 +177,15 @@ public void clusterKMeans(ClusterKMeansParams params,
...
                 .withInputWsObjects(Arrays.asList(params.getInputData()));
-        getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(params.getOutWorkspace())
+
+	WorkspaceClient wsc = getWsClient();
+        List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,Map<String,String>>> getInfo = wsc.saveObjects(new SaveObjectsParams().withWorkspace(params.getOutWorkspace())
                 .withObjects(Arrays.asList(new ObjectSaveData()
                 .withType("KBaseFeatureValues.FeatureClusters").withName(params.getOutClustersetId())
                 .withData(new UObject(toSave)).withProvenance(provenance))));
+
+        String outref = getInfo.get(0).getE7()  + "/" + getInfo.get(0).getE1() + "/" +  getInfo.get(0).getE5();
```
Comment (MrCreosote) [inline lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java]: DRY this up, it's repeated multiple times

**Review by MrCreosote — CHANGES_REQUESTED** [review]
Few things, no big deals I think

File: lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java (hunk 1 above)
Comment (realmarcin) [reply lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java]: done see
https://github.com/kbaseapps/FeatureValues/pull/21/commits/e6322116fba34b1e3a0060a057ed35f65e9e0e15

File: lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java (hunk 2 above)
Comment (realmarcin) [reply lib/src/us/kbase/kbasefeaturevalues/KBaseFeatureValuesImpl.java]: https://github.com/kbaseapps/FeatureValues/pull/21/commits/e6322116fba34b1e3a0060a057ed35f65e9e0e15

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 3]
```diff
@@ -67,6 +67,7 @@
 import us.kbase.auth.ConfigurableAuthService;
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:68]: So looking at these tests, I don't see any changes for the functions that return expression matrices. For the functions that return refs, the only test appears to be that there is some object that exists at that ref, not that the object is as expected.

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: lib/src/us/kbase/kbasefeaturevalues/ExpressionMatrix.java [hunk 4]
```diff
@@ -328,9 +348,79 @@ public void setAdditionalProperties(java.lang.String name, Object value) {
         this.additionalProperties.put(name, value);
     }
 
+    @Override
```
Comment (MrCreosote) [inline lib/src/us/kbase/kbasefeaturevalues/ExpressionMatrix.java:352]: For future reference, eclipse (and probably other IDEs) can generate equals and hashcode for you

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 5]
```diff
@@ -195,38 +197,46 @@ public void testMainPipeline() throws Exception {
...
         Assert.assertEquals(3, k);
         Assert.assertEquals(5, estKRes.getEstimateClusterSizes().size());
         for (int i = 0; i < estKRes.getEstimateClusterSizes().size(); i++) {
             Tuple2 <Long, Double> item = estKRes.getEstimateClusterSizes().get(i);
             Assert.assertEquals(2L + i, (long)item.getE1());
             Assert.assertTrue((double)item.getE2() > 0);
         }
-        impl.estimateKNew(new EstimateKParamsNew().withInputMatrix(testWsName + "/" +
+
+	List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,Map<String,String>>> info2 = getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName)
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: I don't understand this. You run the method and get the return result. You then save that result to the workspace, pull it back, and then compare that object to the object returned by the impl result. I'm not sure what this tests.

I would expect that you manually construct an expected result for `estKRes` and then compare `extKRes` to that expected result.

Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Also you're only checking a portion of the returned object instead of the entire object, it seems.

Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: If the method also saves the result to the workspace as well as returning it, then the test should pull that result, not save a new object.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 6]
```diff
@@ -258,11 +268,26 @@ public void testMainPipeline() throws Exception {
         Assert.assertEquals(2, clSet3.getFeatureClusters().get(1).getIdToPos().size());
         Assert.assertTrue(clSet3.getFeatureDendrogram().startsWith("("));
         Assert.assertTrue(clSet3.getFeatureDendrogram().endsWith(");"));
+
+
+	List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,Map<String,String>>> info = getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName)
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: This is another case of saving an object to the ws, getting it, and then comparing what you saved to what you got. It doesn't seem to test feature values code at all.

There's more places like this in the code, but I'm going to stop at 2 comments.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 7]
```diff
@@ -329,38 +356,87 @@ public void testPyScikitKMeans() throws Exception {
...
-        ObjectData res = wscl.getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+        ObjectData res = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
                 .withName(clustObj1Name))).get(0);
+
         FeatureClusters clSet = res.getData().asClassInstance(FeatureClusters.class);
         checkKMeansForSample(clSet);
     }
-    
+
+    @Test
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:414]: I don't understand the point of this test at all. It's just testing the workspace, not feature values.

Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:414]: I think there's some misunderstanding here about testing, so before I spend more time on the review I think we should have a bit of a natter and get on the same page, to use a ghastly business idiom.

**Review by MrCreosote — CHANGES_REQUESTED** [review]

**Comment by MrCreosote** [comment]
Note to self: 
2 methods now return an ExpressionMatrix
6 methods now return a ref

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 8]
```diff
@@ -195,61 +197,59 @@ public void testMainPipeline() throws Exception {
...
-        //Assert.assertNotNull("exprObjName ",testWsName + "/" + exprObjName);
-        //Assert.assertEquals("exprObjName ",testWsName + "/" + exprObjName);
-        //Assert.assertNotNull("estimObjName ", estimObjName);
-        //
+                        .withData(new UObject(data)))));
         /////////////// estimate K /////////////////
         EstimateKParams ekp = new EstimateKParams().withInputMatrix(testWsName + "/" +
                 exprObjName).withOutWorkspace(testWsName).withOutEstimateResult(estimObjName);
-        impl.estimateK(ekp, token, getContext());
-        ObjectData res1 = wscl.getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: Why are these lines deleted? I assume these are checking that the object saved by the method is correct? That's still necessary, right?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 9]
```diff
@@ -195,61 +197,59 @@ public void testMainPipeline() throws Exception {
...
-                .withName(estimObjName))).get(0);
-        EstimateKResult estKRes = res1.getData().asClassInstance(EstimateKResult.class);
+        EstimateKResult estKRes = impl.estimateK(ekp, token, getContext());
+
+        //ObjectData res1 = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+        //        .withName(estimObjName))).get(0);
+        //EstimateKResult estKRes = res1.getData().asClassInstance(EstimateKResult.class);
         long k = estKRes.getBestK();
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: Is there a reason the tests only check portions of the object?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: I see there's more cases of this, I won't point them all out.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: I also see there's more cases of this, I won't point them all out.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 10]
```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
...
                 .withName(clustObj1Name))).get(0);
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(getInfoEKN.get(0).getE7()  + "/" + getInfoEKN.get(0).getE1() + "/" +  getInfoEKN.get(0).getE5(), getId);
+
         FeatureClusters clSet = res.getData().asClassInstance(FeatureClusters.class);
         checkKMeansForSample(clSet);
     }
-    
-    @Test
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:345]: Is this supposed to be deleted? That'll deactivate the test.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 11]
```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
...
+
+        //test matrix imputation
         Assert.assertEquals(0, getNullCount(matrix.getData()));
-        Assert.assertEquals(0.325, (double)matrix.getData().getValues().get(0).get(0), 1e-10);
+        Assert.assertEquals(0.325, (double) matrix.getData().getValues().get(0).get(0), 1e-10);
+
+        //reload with returned object id
+        ExpressionMatrix res = loadfromWs(getId).asClassInstance(ExpressionMatrix.class);
+        //reset from null in data before imputation
+        data.getData().getValues().get(0).set(0, res.getData().getValues().get(0).get(0));
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Can you explain what's going on in this line? Maybe expand the comment. It's not clear to me.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 12]
```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
...
+        Assert.assertEquals(data.getData().toString(), matrix.getData().toString());
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: This appears to be testing the input matrix against the result matrix from the method. Shouldn't they be different? My assumption is that the method is not a noop, but maybe I'm wrong.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 13]
```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
...
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(infoCM.get(0).getE7()  + "/" + infoCM.get(0).getE1() + "/" +  infoCM.get(0).getE5(), getId);
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Why is the method returning the ref for an object that was created prior to the method being run? Doesn't the method do anything?

Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: There's more cases of this, I won't mark them

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 5 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: indeed, there were a mix of cases for the returns -- this test only applied to returns which were a string ref. removed now.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: this is deleted because it is now superseded by the test based on the object from the impl method return

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: for EstimateKResult there are only two relevant fields and those have always been tested

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 10 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:345]: thanks, fixed

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: this is testing the matrix values globally, but have to change one in test matrix to account for data imputation

actually this is already covered under testSaveMatrix() -- but can't hurt

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 12 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: they should be different and one value is reset to make the comparison equal, see comments above

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 13 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: this is now fixed -- the object id has to be incremented in this case

Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: and the few other cases

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: I don't understand this. AIUI, the app saves the matrix to the workspace. The changes here make the app also return the matrix at the end of the function. Testing the contents of the matrix returned by the function does not test that the matrix was saved to the ws correctly, or even at all.

Or do I not understand what the function is doing?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: Ok, it'd be good to document that. To a person who is a non-expert in the codebase it looks like parts of the data are ignored for no reason.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: So how does the test ensure that the imputed value is correct?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 12 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Ok, so the only difference is the imputed value(s) and those are set equal to the output matrix in the input matrix, and then the matrices are compared to ensure the method has done nothing other than add the imputed values?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 13 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Now I'm confused as to how the tests passed previously - the method behavior didn't change at all AFAICT but the tests passed, but now the tests add 1 to the object ID or version and the tests still pass. How is that possible?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: wouldn't they just look in the object spec?
I added a comment line ...

is it standard practice to test null fields?
I added the only other possible test, for the empty additionalProperties

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 12 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: that's right, with the exception that the actual imputed value itself is not tested -- will add.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: this test does not but a few lines above:

Assert.assertEquals(0.325, (double) matrix.getData().getValues().get(0).get(0), 1e-10);

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 13 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: ha, it turns out that these tests did not pass as I think the classes got cached at some point along the way ... but pass now with the +1 increments.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: > wouldn't they just look in the object spec?

I doubt it. I didn't. I'd think most people would just go 'why the heck is the object only partially tested?'

Even if they did look at the spec, a non-expert in this module would also probably not know which fields in the spec were important and which weren't.

> is it standard practice to test null fields?

I do *shrug*. If the field is supposed to be null and not some other value, that should be tested.

Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: > I added the only other possible test, for the empty additionalProperties

Oh, that I wouldn't bother about unless there's something you're expecting to be in there.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Ok, i'm a little confused then - the test pulls the same object from the workspace twice, and runs different tests on the two instances of the same object?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 12 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: I assume these haven't been added yet? I don't see anything in the subsequent commits.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: The change is that it returns an EstimateKResult. What is being tested are the fields in this returned object, for both of the estimateK methods. No matrices here BTW ...

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: based on your comments above I'm keeping the additionalProperties test ie "If the field is supposed to be null and not some other value, that should be tested."

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Yes, but it pulls the object using two different ways 1) ws and object name specified in the code and 2) ws reference returned by the method. It may be confusing at first but I we do want to test both the returned object contents as well as the ws ref for that object.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 12 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: Right, added now and in new PR:
https://github.com/realmarcin/FeatureValues/blob/5e82fbce8a75508aa8be1ad2250d18f2678039e3/test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java#L398

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: > The change is that it returns an EstimateKResult. 

I understand that, but it's still saving the EstimateKResult, right? So why are the tests for that deleted?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 9 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:208]: Probably not necessary IMO, but doesn't hurt for sure.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (MrCreosote) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: >  It may be confusing at first but I we do want to test both the returned object contents as well as the ws ref for that object.

But you can do that by just pulling once via the returned ref, surely?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: because testing the saved vs the returned object seems redundant per discussions above?

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 11 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java]: sure, you could pull once but then test is combined, isn't it better to isolate failures?

but I just combined into one load...

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java (hunk 8 above)
Comment (realmarcin) [reply test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:209]: ok, well the object created in the code vs saved in the workspace does make sense. I replicated those tests for the two estimateK methods in the new PR.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 14]
```diff
@@ -195,61 +197,111 @@ public void testMainPipeline() throws Exception {
...
-            Assert.assertEquals((double)entry.getE2(), (double)entryNew.getE2(), 1e-10);
+            Assert.assertEquals((long) entry.getE1(), (long) entryNew.getE1());
+            Assert.assertEquals((double) entry.getE2(), (double) entryNew.getE2(), 1e-10);
+        }
+
+        //loading saved result object
+        ObjectData res1New = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+                .withName(estimObjName))).get(0);
+        EstimateKResult estKNewResLoad = res1New.getData().asClassInstance(EstimateKResult.class);
+        long kNewLoad = estKNewResLoad.getBestK();
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:272]: It seems like there could be some DRYing up of these repeated test methods in the future

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 15]
```diff
@@ -195,61 +197,111 @@ public void testMainPipeline() throws Exception {
...
+        Assert.assertNotNull("k exists ", kLoad);
+        Assert.assertNotNull("c size exists ", estKResLoad.getEstimateClusterSizes().size());
+        Assert.assertEquals(3, kLoad);
+        Assert.assertEquals(5, estKResLoad.getEstimateClusterSizes().size());
+        Map<String, Object> additionalPropertiesLoad = new HashMap<String, Object>();
+        Assert.assertEquals(additionalPropertiesLoad, estKRes.getAdditionalProperties());
+        for (int i = 0; i < estKResLoad.getEstimateClusterSizes().size(); i++) {
+            Tuple2<Long, Double> item = estKResLoad.getEstimateClusterSizes().get(i);
+            Assert.assertEquals(2L + i, (long) item.getE1());
+            Assert.assertTrue((double) item.getE2() > 0);
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:244]: Asserting that a value is > 0 doesn't seem like a very stringent test, but it's exactly what the test was previously, so it's no worse than pre-PR anyway.

File: test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java [hunk 16]
```diff
@@ -329,70 +395,90 @@ public void testPyScikitKMeans() throws Exception {
...
+                        .withData(new UObject(data)))));
+        String getId = impl.clusterKMeans(new ClusterKMeansParams().withInputData(testWsName + "/" +
                 exprObjName).withK(3L).withOutWorkspace(testWsName).withOutClustersetId(clustObj1Name),
                 token, getContext());
-        ObjectData res = wscl.getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+        ObjectData res = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
                 .withName(clustObj1Name))).get(0);
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(getInfoEKN.get(0).getE7()  + "/" + (getInfoEKN.get(0).getE1()+1) + "/" +  getInfoEKN.get(0).getE5(), getId);
```
Comment (MrCreosote) [inline test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java:408]: It's somewhat confusing that in some cases the tests check the ref from an object info tuple vs the ref returned by the method, and some tests increment either the object ID or version by 1. If possible I'd choose one or the other.

**Review by MrCreosote — APPROVED** [review]
A few minor things that I don't think need to be corrected right now, but noted for possible future improvements. Going to run the tests locally and then this should be G2G

**Comment by MrCreosote** [comment]
Ugh, forgot this takes a year and a half to build

**Comment by MrCreosote** [comment]
```
    [junit] Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 49.885 sec
    [junit] Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 49.885 sec
    [junit] ------------- Standard Output ---------------
    [junit] Test workspace [test_feature_values_d3c6a9ebbf02_1594350087847] was deleted
    [junit] ------------- ---------------- ---------------
    [junit] 
    [junit] Testcase: testUploader1 took 25.448 sec
    [junit] Testcase: testUploader2 took 0.022 sec
    [junit] Testcase: testUploaderOthers took 0.026 sec
    [junit] Testcase: testMappingToAliases took 23.385 sec
    [junit] 	Caused an ERROR
    [junit] Object #1, Escherichia_coli_str_K-12_substr_MG1655.genome has invalid reference: No read access to id 6308/3/2: Object 3 cannot be accessed: Workspace 6308 is deleted at /ontology_events/0/ontology_ref
    [junit] us.kbase.common.service.ServerException: Object #1, Escherichia_coli_str_K-12_substr_MG1655.genome has invalid reference: No read access to id 6308/3/2: Object 3 cannot be accessed: Workspace 6308 is deleted at /ontology_events/0/ontology_ref
    [junit] 	at us.kbase.common.service.JsonClientCaller.jsonrpcCallStatic(JsonClientCaller.java:411)
    [junit] 	at us.kbase.common.service.JsonClientCaller.jsonrpcCall(JsonClientCaller.java:329)
    [junit] 	at us.kbase.common.service.JsonClientCaller.jsonrpcCall(JsonClientCaller.java:298)
    [junit] 	at us.kbase.workspace.WorkspaceClient.saveObjects(WorkspaceClient.java:487)
    [junit] 	at us.kbase.kbasefeaturevalues.transform.test.ExpressionUploaderTest.testMappingToAliases(ExpressionUploaderTest.java:190)
    [junit] 
```

**Comment by MrCreosote** [comment]
Tests pass on CI, so it looks like tests are bound to CI. That should be fixed at some point
